File: 1944-number-of-visible-people-in-a-queue/1944-number-of-visible-people-in-a-queue.py
class Solution:
    def canSeePersonsCount(self, heights: List[int]) -> List[int]:
        stack = []

        n = len(heights)

        res = [0] * n
        stack.append(heights[-1])

        for i in reversed(range(n-1)):
            if stack[-1] > heights[i]:
                res[i] = 1; 
            else:
                popCount = 0
                while len(stack) > 0 and stack[-1] < heights[i]:
                    stack.pop()
                    popCount += 1
                if stack:
                    popCount += 1
                res[i] = popCount
            
            stack.append(heights[i])

        
        return res


# 10,6,8,5,11,9
# i

# res: 3 1 2 1 1 0

# stack: 10 11

# if stk.top() > cur
#     res = 1

# else

# pop all smaller
#     res = popCount + (stk.size() > 1)

# push onto stack

File: 1944-number-of-visible-people-in-a-queue/1944-number-of-visible-people-in-a-queue.cpp
class Solution {
public:
    vector<int> canSeePersonsCount(vector<int>& heights) {
        int n = heights.size();
        vector<int> res(n);
        stack<int> stk;

        stk.push(heights[n-1]);

        for (int i = n-2; i >= 0; i --) {
            int h = heights[i];
            int popCount = 0;
            while (!stk.empty() && stk.top() <= h) {
                stk.pop();
                popCount ++;
            }

            if (!stk.empty()) {
                popCount ++;
            }

            res[i] = popCount;
            stk.push(h);
        }

        return res;
    }
};

// 10,6,8,5,11,9
//  x

// stack: 11 10

// pop all smaller or equal
// push to queue

// popCount + !stk.empty()

// 3 1 2 1 1 0